File: 1192.critical-connections-in-a-network.py
#
# @lc app=leetcode id=1192 lang=python3
#
# [1192] Critical Connections in a Network
#
from collections import defaultdict
# @lc code=start

# t:O(V+E), s:O(V+E)

class Solution:
    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:
        connect = defaultdict(list)
        for con in connections:
            connect[con[0]].append(con[1])
            connect[con[1]].append(con[0])
        parent = [-1]*n
        seen_time = [-1]*n
        lowest = [None]*n
        res = []
        self.time_count = 0

        def findBridge(node):
            seen_time[node] = self.time_count
            lowest[node] = self.time_count
            self.time_count += 1
            for child in connect[node]:
                if seen_time[child] == -1:
                    parent[child] = node
                    findBridge(child)
                    lowest[node] = min(lowest[node], lowest[child])
                    if lowest[child] > seen_time[node]:
                        res.append([node, child])
                elif child != parent[node]:
                    lowest[node] = min(lowest[node], seen_time[child])

        for i in range(n):
            if seen_time[i] == -1:
                findBridge(i)
        return res


# @lc code=end

File: cpp/1192.critical-connections-in-a-network.cpp
/*
 * @lc app=leetcode id=1192 lang=cpp
 *
 * [1192] Critical Connections in a Network
 */
#include <vector>
#include <functional>
#include <algorithm>
using namespace std;
// @lc code=start
// t:O(V+E), s:O(V+E)
class Solution
{
    int time = 0;

public:
    vector<vector<int>> criticalConnections(int n, vector<vector<int>> &connections)
    {
        unordered_map<int, vector<int>> connect;
        for (auto &con : connections)
        {
            connect[con[0]].push_back(con[1]);
            connect[con[1]].push_back(con[0]);
        }
        vector<int> parent(n, -1);
        vector<int> seen_time(n, -1);
        vector<int> lowest(n);
        vector<vector<int>> res;
        function<void(int)> findBridge = [&](int node)
        {
            seen_time[node] = ++time;
            lowest[node] = time;
            for (auto child : connect[node])
            {
                if (seen_time[child] == -1)
                {
                    parent[child] = node;
                    findBridge(child);
                    lowest[node] = min(lowest[node], lowest[child]);
                    if (lowest[child] > seen_time[node])
                    {
                        res.push_back({node, child});
                    }
                }
                else if (child != parent[node])
                {
                    lowest[node] = min(lowest[node], seen_time[child]);
                }
            }
        };
        for (int i = 0; i != n; ++i)
        {
            if (seen_time[i] == -1)
            {
                findBridge(i);
            }
        }
        return res;
    }
};
// @lc code=end
